dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/ApacheHttpClientBlockingChannel.java

```diff
@@ -455,37 +475,73 @@ public void close() {}
 
     private static final class ResponseInputStream extends FilterInputStream {
 
-        // Client reference is used to prevent premature termination
-        @Nullable
-        private ApacheHttpClientChannels.CloseableClient client;
-
         private final HttpClientResponse response;
 
-        ResponseInputStream(
-                @Nullable ApacheHttpClientChannels.CloseableClient client,
-                InputStream stream,
-                HttpClientResponse response) {
+        ResponseInputStream(InputStream stream, HttpClientResponse response) {
             super(stream);
-            this.client = client;
             this.response = response;
         }
 
         @Override
-        public void close() throws IOException {
-            try {
-                try {
-                    response.close();
-                } finally {
-                    super.close();
-                }
-            } finally {
-                client = null;
+        public int read() throws IOException {
+            checkOpen();
+            return super.read();
+        }
+
+        @Override
+        public int read(byte[] buffer) throws IOException {
+            checkOpen();
+            return super.read(buffer);
+        }
+
+        @Override
+        public int read(byte[] buffer, int off, int len) throws IOException {
+            checkOpen();
+            return super.read(buffer, off, len);
+        }
+
+        @Override
+        public long skip(long num) throws IOException {
+            checkOpen();
+            return super.skip(num);
+        }
+
+        @Override
+        public void close() {
+            if (response.isOpen()) {
+                response.close();
+                // no need to close the delegate stream itself, closing the response is sufficient
```

Comment:
Is that always the case? what if they have an implementation that like holds onto a pooled bytebuffer? does calling close on the response guarantee to close any opened input streams?

Reply:
Yep, the client doesn’t pool anything beyond connections, and buffers are created with connections (and collected when the connection is closed)